diff --git a/SPECS/glibc.spec b/SPECS/glibc.spec
index 9bd07c9..c389711 100644
--- a/SPECS/glibc.spec
+++ b/SPECS/glibc.spec
@@ -1,6 +1,6 @@
 %define glibcsrcdir glibc-2.12-2-gc4ccff1
 %define glibcversion 2.12
-%define glibcrelease 1.212%{?dist}.3
+%define glibcrelease 1.212.1%{?dist}.3
 %define run_glibc_tests 1
 %define auxarches athlon sparcv9v sparc64v alphaev6
 %define xenarches i686 athlon
@@ -283,6 +283,7 @@
 Patch251: glibc-rh1555930-1.patch
 Patch252: glibc-rh1555930-2.patch
 Patch253: glibc-rh1577437.patch
+Patch254: remove-vsyscall.patch

 Buildroot: %{_tmppath}/%{name}-%{version}-%{release}-root-%(%{__id_u} -n)
 Obsoletes: glibc-profile < 2.4
@@ -739,6 +739,7 @@
 %patch251 -p1
 %patch252 -p1
 %patch253 -p1
+%patch254 -E -p3

 # A lot of programs still misuse memcpy when they have to use
 # memmove. The memcpy implementation below is not tolerant at

